grype/matcher/apk/matcher.go

```diff
@@ -23,39 +26,115 @@ func (m *Matcher) Type() match.MatcherType {
-
 	// find Alpine SecDB matches for the given package name and version
 	secDbMatches, err := common.FindMatchesByPackageDistro(store, d, p, m.Type())
 	if err != nil {
 		return nil, err
 	}
 
-	for _, secDbMatch := range secDbMatches {
-		secDbCandidates[secDbMatch.Vulnerability.ID] = append(secDbCandidates[secDbMatch.Vulnerability.ID], secDbMatch)
+	cpeMatches, err := m.cpeMatchesWithoutSecDbFixes(store, d, p)
+	if err != nil {
+		return nil, err
 	}
 
-	// find NVD matches specific to the given package name and version
-	var cpeCandidates = make(map[string][]match.Match)
+	// keep all secdb matches, as this is an authoritative source
+	matches = append(matches, secDbMatches...)
+
+	// keep only unique CPE matches
+	matches = append(matches, deduplicateMatches(secDbMatches, cpeMatches)...)
+
+	return matches, nil
+}
+
+func (m *Matcher) cpeMatchesWithoutSecDbFixes(store vulnerability.Provider, d *distro.Distro, p pkg.Package) ([]match.Match, error) {
+	// find CPE-indexed vulnerability matches specific to the given package name and version
 	cpeMatches, err := common.FindMatchesByPackageCPE(store, p, m.Type())
 	if err != nil {
 		return nil, err
 	}
 
-	for _, cpeMatch := range cpeMatches {
-		cpeCandidates[cpeMatch.Vulnerability.ID] = append(cpeCandidates[cpeMatch.Vulnerability.ID], cpeMatch)
+	cpeMatchesByID := matchesByID(cpeMatches)
+
+	// remove cpe matches where there is an entry in the secDB for the particular package-vulnerability pairing, and the
+	// installed package version is >= the fixed in version for the secDB record.
+	secDbVulnerabilities, err := store.GetByDistro(*d, p)
+	if err != nil {
+		return nil, err
+	}
+
+	secDbVulnerabilitiesByID := vulnerabilitiesByID(secDbVulnerabilities)
+
+	verObj, err := version.NewVersionFromPkg(p)
+	if err != nil {
+		return nil, fmt.Errorf("matcher failed to parse version pkg='%s' ver='%s': %w", p.Name, p.Version, err)
 	}
 
-	// package is vulnerable if there is a match in the alpine SecDB and NVD for the same CVE
-	for cve, cpeCandidatesForCve := range cpeCandidates {
-		// by this point all matches have been verified to be vulnerable within the given package version relative to the vulnerability source
-		_, ok := secDbCandidates[cve]
-		if ok {
-			// this is a match, use the NVD records as the primary record source (no need to merge)
-			matches = append(matches, cpeCandidatesForCve...)
+	var finalCpeMatches []match.Match
+
+cveLoop:
+	for id, cpeMatchesForID := range cpeMatchesByID {
+		// check to see if there is a secdb entry for this ID (CVE)
+		secDbVulnerabilitiesForID, exists := secDbVulnerabilitiesByID[id]
+		if !exists {
+			// does not exist in secdb, so the CPE record(s) should be added to the final results
+			finalCpeMatches = append(finalCpeMatches, cpeMatchesForID...)
+			continue
+		}
+
+		// there is a secdb entry...
+		for _, vuln := range secDbVulnerabilitiesForID {
+			// ...is there a fixed in entry? (should always be yes)
```

Comment:
I think this implements the logic, but seems redundant to me. The SecDB match in Match() should catch that case and if there isn't a match from SecDB then the NVD match should be ignored. What case does this catch that isn't caught by the earlier secDB matching pass?

Reply:
technically yes, this is redundant. However, since we're pulling out entries from the DB via `GetByDistro` (instead of using any earlier results) we are making no assumptions about the vulnerabilities that are returned. This means also checking explicitly that the package is vulnerable for the specific vulnerability details returned... which means we should check the version constraint.

The same is true for a list of vulnerabilities. We know that since SecDB is the only backing dataset, there should only be one vulnerability for each ID for each package, not multiple. However, the query makes no assumptions there, so we treat the return as if multiple could be returned, even though it shouldn't happen under the current circumstances.